### DISCUSSION TITLE
requirement of merging local repos?

### DISCUSSION THREAD

**Discussion by dmastrop** [community/community#51995] [Programming Help] [answered]
### Body

hi all,

I was wondering if this should be normal procedure in making sure all branches are the same at both local and remotes.

_Should the locals always have the same 
show log --pretty=oneline ?_

This is what happens after I do a remote merge of the master and feature branch, every time, even though I do a pull on the locals after the remote merging.

(Formatting is a bit off but I hope this is clear.)

I end up with one commit off in one branch. See below:

**This is the feature branch: (it is short one commit relative to master, see below)**
git log --pretty=oneline
92d9896f711b0c56bef857958010d4c0f94cc1b7 (HEAD -> feature, origin/feature) commit App.js latest change to feature branch
868abeba4f7bfdda93c771e5cd3d52fbf9de7b36 dummy commit to feature comments deploy.yml

**This is the master branch:**
git log --pretty=oneline
28ecb3bdbebdf646c0ab993514e7d951577d42c7 (HEAD -> master, origin/master) Merge pull request #9 from dmastrop/feature
92d9896f711b0c56bef857958010d4c0f94cc1b7 (origin/feature, feature) commit App.js latest change to feature branch

=====

note: show remote origin master and feature show everything is up to date on remotes relative to locals

git remote show origin
* remote origin
  Fetch URL: https://github.com/dmastrop/docker-react.git
  Push  URL: https://github.com/dmastrop/docker-react.git
  HEAD branch: master
  Remote branches:
    feature tracked
    master  tracked
  Local refs configured for 'git push':
    feature pushes to feature (up to date)
    master  pushes to master  (up to date)

==========

On master looks ok as expected based on the commit logs:

 git branch --merged
  feature
* master

**I see this on feature branch:**

 git branch --merged
* feature

 git branch --no-merged   **<<< need to merge feature <—— master   Checkout feature and do a git merge master**
  master

=========

I need to merge feature <----- master.  Checkout the feature and do a "git merge master"

git checkout feature
Switched to branch 'feature'

 git status
On branch feature
nothing to commit, working tree clean

 git merge master
Updating 92d9896..28ecb3b
Fast-forward

This seems to have addressed the issue:

git branch --merged
* feature
  master

 git branch --no-merged
<no output as expected>

==========

**commit logs look good** 

master:
 git log --pretty=oneline
28ecb3bdbebdf646c0ab993514e7d951577d42c7 (HEAD -> master, origin/master, feature) Merge pull request #9 from dmastrop/feature
92d9896f711b0c56bef857958010d4c0f94cc1b7 (origin/feature) commit App.js latest change to feature branch

feature:
git log --pretty=oneline
28ecb3bdbebdf646c0ab993514e7d951577d42c7 (HEAD -> feature, origin/master, master) Merge pull request #9 from dmastrop/feature
92d9896f711b0c56bef857958010d4c0f94cc1b7 (origin/feature) commit App.js latest change to feature branch

======

**now the remotes are off:**

 git checkout feature
Already on 'feature'
git remote show origin
* remote origin
  Fetch URL: https://github.com/dmastrop/docker-react.git
  Push  URL: https://github.com/dmastrop/docker-react.git
  HEAD branch: master
  Remote branches:
    feature tracked
    master  tracked
  Local refs configured for 'git push':
    ****feature pushes to feature (fast-forwardable)****
    master  pushes to master  (up to date)

git push origin feature
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
To https://github.com/dmastrop/docker-react.git
   92d9896..28ecb3b  feature -> feature

git remote show origin
* remote origin
  Fetch URL: https://github.com/dmastrop/docker-react.git
  Push  URL: https://github.com/dmastrop/docker-react.git
  HEAD branch: master
  Remote branches:
    feature tracked
    master  tracked
  Local refs configured for 'git push':
    **feature pushes to feature (up to date)
    master  pushes to master  (up to date)**

master is ok:

 git checkout master
Switched to branch 'master'
 git remote show origin
* remote origin
  Fetch URL: https://github.com/dmastrop/docker-react.git
  Push  URL: https://github.com/dmastrop/docker-react.git
  HEAD branch: master
  Remote branches:
    feature tracked
    master  tracked
  Local refs configured for 'git push':
    feature pushes to feature (up to date)
    master  pushes to master  (up to date)

=======

my question is: _this seems circular and tedious in some way. Is there a better way to get locals and remotes all on the same state relative to commits?_

There were no code changes just a merge done remotely (after a code change) that causes this.

After merging remotely it seems like this procedure always has to be always be done, correct? (to get the local feature and master commits in line with that last commit on the merge)

_It does not seem like leaving in the former state is bad in any way (in terms of synch) because on the next real commit (code change, etc), everything will get updated, correct?_

warm regards

### I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

- [X] Yes

**Comment by airtower-luna** [comment, marked as answer]
> There were no code changes just a merge done remotely (after a code change) that causes this.

You did a merge with a merge commit, that's why there's an additional commit on `master`. That's entirely normal.

The common approach would be to *delete* (or just leave) the `feature` branch after the merge (`git branch -d feature`), and create a new feature branch (starting from `master` again) for the next feature you're working on. If you want to keep working from the same branch for some reason doing a local fast-forward merge from `master` as you did is exactly the right thing to do.
